overlord/snapstate/snapstate.go

```diff
@@ -719,6 +719,28 @@ func Update(st *state.State, name, channel string, revision snap.Revision, userI
 	if err != nil {
 		return nil, err
 	}
+
+	// see if we need to update the channel
+	if infoErr != nil && snapst.Channel != channel {
```

Comment:
Good point, infoErr can only be NoUpdateAvailableError because otherwise we would have exited earlier. However this is very much non-obvious so I will modify the code to make that clearer.